### ISSUE TITLE
Allow for triggering GUI tooltips on focus

### ISSUE THREAD

**Issue by Serenitor** [godotengine/godot-proposals#2584]
### Describe the project you are working on

A game with a settings screen where I want to describe certain settings in tooltips - and not just when the player is navigating via mouse.

### Describe the problem or limitation you are having in your project

Currently, tooltips - which can be specified via the `hint_tooltip` property of `Control` - are only shown when hovering the node with your mouse cursor. This means that when navigating the GUI via keyboard or gamepad, they never appear. This is unfortunate since of course, when navigating with these devices, tooltips or no less relevant or useful.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I suggest to add a `trigger_on_focus` boolean property to the HInt section of Control.
When enabled, the tooltip would also appear when the Control node receives focus, and disappear the moment it loses focus.
This could take the project setting `gui/timers/tooltip_delay_sec` into account as well, personally I would want these on-focus tooltips to appear immediately though.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I believe no mock-ups are necessary as it is a very straight-forward proposal. Let me know if anything remains unclear.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Anyone who wants to give additional explanation to GUI options will consider making use of this feature. The alternative would be to implement your own tooltip system that also works with keyboard and gamepad navigation, but that would certainly require more than a few lines, and it would be a rather poor experience to have to implement something that engine technically already provides.

### Is there a reason why this should be core and not an add-on in the asset library?

As the existing tooltip system is already part of core, I think extending it directly makes most sense.

**Comment by Serenitor** [comment]
As a bonus, it would be brilliant if on-focus tooltips would update their text in real-time when `hint_tooltip` is changed via script.
Then, `Slider` nodes would become more powerful, as they could display their current value in the tooltip while the player moves the slider.

Concrete use-case: A setting to limit FPS. `Slider` is definitely a more appropriate choice than`LineEdit` or `PopupMenu` here, but of course the player needs to understand what the current slider position means in terms of specific FPS.

**Comment by Calinou** [comment]
To be fair, I think you should implement your own tooltip system when you need complex features like:

- Formatting in tooltips.
- Positioning in a fixed location, or a more convenient location than the one automatically chosen.
- Fade transitions.
- Real-time updates (although this could be implemented in the built-in tooltip system).

The built-in tooltip system is designed for non-game applications like the editor. I wouldn't expect it to become a polished system for most games.

**Comment by Serenitor** [comment]
Right, I actually don't *need* any of those complex features though. :)
I think I am proposing very basic functionality here: Making the existing tooltips system usable for keyboard and gamepad navigation. So it's hardly even a new feature - it's merely about making it usable.

I guess I shouldn't have brought up the "as a bonus" application. I am aware that I *could* implement tooltips in all sorts of ways myself. However I think it's a waste to not leverage the existing system and that it's bloat to build my own.

I'm surprised about you disregarding tooltips for games. Are tooltips not a common feature used in games, and aren't there many very UI heavy games?

**Comment by Calinou** [comment]
> I'm surprised about you disregarding tooltips for games. Are tooltips not a common feature used in games, and aren't there many very UI heavy games?

The thing is that there's not "one true way" to do tooltips in games. Some games such as Minecraft will display tooltips for inventory items that move along the cursor. Some games will use "tooltips" displayed at the bottom of the screen to describe various options in a graphic settings menu.

**Comment by Serenitor** [comment]
Makes sense.
For non-floating tooltips, I agree that people should build their own thing.
But I'd say the floating nature of tooltips is exactly what makes them so useful in most applications, and that's already implemented - ready to use... if your input device of choice is a mouse, that is.

Anyway, I am strictly sticking to the trigger on focus idea for this proposal now, and no extras.
Tooltips would appear aligned to the bottom left corner of the Control node, I am thinking.

**Comment by vvvvvvitor** [comment]
Currently I also need this feature. Applications would also benefit from this being added, as many applications support keyboard alongside with the mouse. I have no idea why this feature only works when hovering.

**Comment by Calinou** [comment]
> I have no idea why this feature only works when hovering.

This is standard behavior in applications. Tooltips only appear on hover, not on focus.

**Comment by h0lley** [comment]
I don't think they were suggesting that it's expected for tools and websites to have tooltips appear on focus.

when an user opts for keyboard or gamepad as their input device, or if those are the only available input devices, then there's no good reason for tooltips to no longer be available. With gamepad and keyboard navigation, too, tooltips are just as relevant and useful.

And when Godot users would merely need to tick a checkbox to enable tooltip support for gamepad or keyboard navigation instead of having to write an entire system then I don't see why that would't be a feature many projects would make use of.

the highly customizable tooltip feature is already fully built-in - merely the trigger options are limiting. pretty unfortunate.

**Comment by metamagicshaman** [comment]
I assumed showing tooltips via gamepad was a no-brainer and already implemented. I'm honestly shocked there's even a debate about whether it should be a thing or not.

**Comment by Okxa** [comment]
Background:  
I am developing a plugin in which the user can switch between several modes, and I wanted to add a button with a question mark icon which would show some info about the selected mode. I thought to save some time and add the info as the `tooltip_text` in the button, and it works as expected, but I also wanted to show the tooltip on button press, in case it's not immediately obvious to the user what the buttons purpose is. But as it is not possible, I would have to create a new control that is show on button press (and also on hover).

**Anyways**, if there was a function to show a tooltip for any control at any position, I think the use cases presented in this proposal would be possible to implement too.

The mouse hover tool tip is shown with `Viewport::_gui_show_tooltip()`, which is called from `Viewport::_gui_input_event`, where it finds the control under mouse and sets some values for the tooltip:
```c++
if (!is_tooltip_shown && over->can_process()) {
    Vector2 new_tooltip_pos = over->get_screen_transform().xform(pos);
    if (over != gui.tooltip_control || gui.tooltip_pos.distance_squared_to(new_tooltip_pos) > 25) {
        if (gui.tooltip_timer.is_valid()) {
            gui.tooltip_timer->release_connections();
        }
        gui.tooltip_control = over;
        gui.tooltip_pos = new_tooltip_pos;
        gui.tooltip_timer = get_tree()->create_timer(gui.tooltip_delay);
        gui.tooltip_timer->set_ignore_time_scale(true);
        gui.tooltip_timer->connect("timeout", callable_mp(this, &Viewport::_gui_show_tooltip));
    }
}
```
If I read the code correclty, maybe it could be possible (but not sure how much work) to implement a method something like `Viewport.gui_show_tooltip_at_position(control:Control, global_position:Vector2, delay_timer:float)` which would set the above parameters and call `Viewport::_gui_show_tooltip()` after the delay.

Although that still would require to get a reference to the current viewport\*, it would allow showing a tooltip on demand for any control.

\**which in EditorPlugins case would require parsing editor node tree to get the root viewport, or a helper method in EditorInterface to either get the viewport or rather show a tooltip*

**Comment by daenvil** [comment]
Currently needing this as well. If at least we had something like ``show_tooltip_at`` and ``hide_tooltip_at`` methods then it could be easily implemented with an addon, but currently I think we would need to reimplement the whole tooltip system just to add this simple feature.

Tooltips are essential in many kinds of videogames and I don't see why Godot's implementation wouldn't allow to use them without a mouse.
